### ISSUE TITLE
callbacks in grpc.cc

### ISSUE THREAD

**Issue by Michaelvll** [google-research/seed_rl#26]
Hi,
I am new in distributed learning. I am wondering what is the 'callback' in the `FnType::operator()` arguments used for in the `grpc.cc`. It seems that the python code only passes 'args' into the function, but there are quite many codes handling the callbacks. Is the 'callback' argument used in seed_rl?

**Comment by lespeholt** [comment]
Callbacks are used on the server side for asynchronous grpc and also internally when calling the registered tensorflow function.


https://github.com/google-research/seed_rl/blob/master/grpc/ops/grpc.cc#L721
https://github.com/google-research/seed_rl/blob/master/grpc/ops/grpc.cc#L716

**Comment by Michaelvll** [comment]
Thank you so much for your quick response! The callback is used for 'timeout' and shape/status checking in `f_callback`, but I am still curious about what the `callback`  could be for asynchronous grpc, i.e. what callback function would be passed into the `operator()` function.

Another quick question, I am wondering if multiple batches will be inferred on a same device in parallel.
https://github.com/google-research/seed_rl/blob/eff7aaa7ab5843547fbf383fcc747b7a8ca67632/agents/vtrace/learner.py#L387-L390
For example, if we only have one inference device and an inference task is running on it when another batch received by the grpc, will the coming batch be assigned to the same inference device and run in parallel with the original task or waits for the former one until it completes?

**Comment by lespeholt** [comment]
I'm not sure I understand the first part of the question so can you rephrase it?

Whenever there is a batch ready the registered function 'inference' will be called, so there is no waiting. I have experimented with adding features to grpc to restrict how many can run at any given time. It's easy to restrict it in TF as well with e.g. tf.CriticalSection

**Comment by Michaelvll** [comment]
Sorry for the implicitness.  What I would like to know is whether there are any callbacks used in the SEED-RL other than the callbacks that appeared in the grpc.cc code.

**Comment by lespeholt** [comment]
No, the only callbacks are in the grpc.cc

**Comment by Michaelvll** [comment]
Thank you so much for your patience! That really helps! I am closing this issue.
